FiveRockingFingers/FRF.Core/Services/ArtifactsService.cs

```
@@ -164,18 +164,19 @@ public async Task<ServiceResponse<Artifact>> Save(Artifact artifact)
                 return new ServiceResponse<Artifact>(new Error(ErrorCodes.ProjectNotExists, $"There is no project with Id = {artifact.ProjectId}"));
             }
 
-            if(! await _dataContext.ArtifactType.AnyAsync(at => at.Id == artifact.ArtifactTypeId))
+            var artifactType = await _dataContext.ArtifactType.Include(at => at.Provider).SingleOrDefaultAsync(at => at.Id == artifact.ArtifactTypeId);
+
+            if (artifactType == null)
             {
                 return new ServiceResponse<Artifact>(new Error(ErrorCodes.ArtifactTypeNotExists, $"There is no artifact type with Id = {artifact.ArtifactTypeId}"));
-            }            
+            }
+
+            artifact.ArtifactType = _mapper.Map<ArtifactType>(artifactType);
```

I'm not quite sure if this is needed or not, but I believe that having the ArtifactTypeId property associated to the correct Id should be enough, although as I said, I'm not sure. (If that's so, that means the previous query to the database could be a little simpler, and not include the Provider)

---

The problem is that  the settings validator uses a different file for XML validation, depending on the artifact type and I think we should check the provider first (as we did with the MapArtifact function in this service) just in case we have two artifact types with the same name and different provider

---

Giving it a second thought maybe we could change both switch (the one in the MapArtifact function and the one in the SettingsValidator) to just check for the ArtifactTypeId. What do you think?

---

It works both ways. If you ask me? I think I prefer to avoid the query in line 167 and just let the db blow up if the provided _ArtifactTypeId_ doesn't exist. Which means... code as simple as possible until complexity is needed.

---

The problem is that we need that query for the settings validation. For now, the settings validation is doing something like:

 ```
switch (artifact.ArtifactType.Provider.Name)
{
    case ArtifactTypes.Custom:
        "Validate Settings with file CustomArtifact.xsd"
        break;
    case ArtifactTypes.Aws:
        switch (artifact.ArtifactType.Name)
        {
            case AwsS3Descriptions.Service:
                 "Validate Settings with file AwsS3.xsd"
                break;
            case AwsEc2Descriptions.Service:
                "Validate Settings with file AwsEc2.xsd"
                break;
        }
}
```
So we need the provider name and the artifact type name.